### DISCUSSION TITLE
Why are ExperimentalWarning flags logged as Errors?

### DISCUSSION THREAD

**Discussion by john-zaprite** [vercel/next.js#44025] [Help]
As per: https://github.com/vercel/next.js/discussions/41124

Why are _ExperimentalWarning_ flags logged as 'Error'? Is this a Vercel/next.js issue? 

This is causing havoc with my error logs being flooded with emails since upgrading to Node v18.

**Comment by icyJoseph** [comment]
Do you mean this?

```shell
(node:22467) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
```

If so, that's a Node.js warning.

You can set `NODE_NO_WARNINGS=1` when running your script, but that's a global flag, and it'll silence other warnings too.

For example: 

```shell
NODE_NO_WARNINGS=1 yarn dev
```

**Reply by john-zaprite** [reply]
Yes, that's the one. But in my logs it is preceded by 'ERROR', which means my emails are flooded with error notifications for every fetch API call:
<img width="1287" alt="ERROR-ExperimentalWarning" src="https://user-images.githubusercontent.com/12265052/207604199-671321c2-ca7f-48bb-8813-43a8f5ab20d8.png">

I'm just trying to figure out why it is marked as an Error when it is not an Error.

Interestingly, in Realtime Logs, it doesn't show in the Errors tab, but just in the 'All' tab. So Vercel is able to conditionally recognize it. But I want to know why it carries the ERROR flag in the first place?

And I suspect it is something to do with `stderr` as per this issue in Node:
https://github.com/nodejs/node/discussions/45820#discussioncomment-4373244

**Reply by icyJoseph** [reply]
The thing is that processes either write to `stdout` or to `stderr`. Warnings are piped to `stderr`, according to Node.js docs anyway.

If we read here: https://vercel.com/docs/concepts/deployments/runtime-logs#level


> - Serverless and Edge logs with an output stdout are marked as Info (for example, console.log)
> - Serverless and Edge logs with an output stderr are marked as Error (for example, console.error)
> - Requests with a status code of 4xx are considered a Warning
> - Requests with a status code of 5xx are considered Error
> - All other requests are considered Info

The problem being that the Node.js documentation says, https://nodejs.org/api/console.html#consolewarndata-args:

> console.warn([data][, ...args])[#](https://nodejs.org/api/console.html#consolewarndata-args)
> Added in: v0.1.100
> data [<any>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types)
> ...args [<any>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types)
> The console.warn() function is an alias for [console.error()](https://nodejs.org/api/console.html#consoleerrordata-args).

So it is writing to `stderr`, at OS level. This is then picked up by the logging system from the stderr pipe, and labeled with the prefix `Error:`. This much we can conclude with all of the info available to us.

**Reply by john-zaprite** [reply]
Yes, this seems to be the same conclusion I have come to. Thanks for the follow-up.

I wonder, if Vercel is already able to conditionally check this, and not show it in the _Error_ tab of _Realtime Logs_, then surely they can manipulate it to show `WARN` instead of `ERROR` before outputting it?

**Reply by icyJoseph** [reply]
I don't know if Vercel can do much about this. If the message comes through `stderr`, then their logging systems emits an Error message. I've seen some logging systems that parse the message for further clues, but that is often also a pitfall.  

Perhaps you can try to put the `NODE_NO_WARNINGS` in their dashboard, or as part of your start script, as to silence those warnings.

**Reply by john-zaprite** [reply]
Right. So I'm guessing they couldn't do the following?
1. Receive the message
2. Parse it and see if it contains 'ExperimentalWarning' text
3. If so, emit as a warning message to their logging system

They are already conditionally checking it—so as not to include it in their realtime Error logs—so I would think they have some control over how this message is parsed/logged in their system.

Hoping someone from @vercel can chime in with a definitive response.

**Reply by icyJoseph** [reply]
Uff that sounds iffy, and for an edge case... Mmm just like the problem you have, that approach opens many other possibilities for chaos. 

Again, either set Node to 16 or silence the warnings. Or ask the Node.js team to give you some fine grained control over which warnings are printed to stderr.
